backend/api_handler.py

```diff
@@ -207,20 +200,60 @@ def handler(event, context):
                 'body': json.dumps(response),
             }
 
-    success, response = graphql_sync(schema=executable_schema, data=query, context_value=app_context)
-
-    dispose_context()
-    response = json.dumps(response)
-
-    log.info('Lambda Response %s', response)
 
-    return {
-        'statusCode': 200 if success else 400,
-        'headers': {
-            'content-type': 'application/json',
-            'Access-Control-Allow-Origin': '*',
-            'Access-Control-Allow-Headers': '*',
-            'Access-Control-Allow-Methods': '*',
-        },
-        'body': response,
-    }
+def validate_and_block_if_maintenance_window(query, groups):
+    # Logic to block when in maintenance
+    # Check if in some maintenance mode
+    # Check if in maintenance status is not INACTIVE
+    # Check if the user belongs to a 'DAAdministrators' group
+    if config.get_property('modules.maintenance.active'):
+        maintenance_mode = MaintenanceService._get_maintenance_window_mode(engine=ENGINE)
+        maintenance_status = MaintenanceService.get_maintenance_window_status(engine=ENGINE).status
+        isAdmin = TenantPolicyValidationService.is_tenant_admin(groups)
+
+        if (
+            (maintenance_mode == MaintenanceModes.NOACCESS.value)
+            and (maintenance_status is not MaintenanceStatus.INACTIVE.value)
+            and not isAdmin
+        ):
+            if query.get('operationName', '') not in MAINTENANCE_ALLOWED_OPERATIONS:
+                return send_unauthorized_response(
+                    query=query,
+                    message='Access Restricted: data.all is currently undergoing maintenance, and your actions are temporarily blocked.',
+                )
+        elif (
+            (maintenance_mode == MaintenanceModes.READONLY.value)
+            and (maintenance_status is not MaintenanceStatus.INACTIVE.value)
+            and not isAdmin
+        ):
+            # If its mutation then block and return
+            if query.get('query', '').split()[0] == 'mutation':
```

Comment:
Lots of issues with using this code. 
I dont like that we are doing a string comparison in the query to match the first word with 'mutation'. It seems super easy to break. For one, if the query string is empty or contains only white spaces, the call to split() will result in an index error because there won't be any elements to access.

The comparison is case-sensitive, so if the query string contains 'Mutation' or 'MUTATION', the condition will not be satisfied.

It assumes that the first word of the query string represents the operation type. However, GraphQL queries can be structured in various ways, and the operation type might not always be the first word.

Use graphql offering OperationDefinitionNode to determine the query operation properly. There are usages of it already in the code.

Reply:
Thanks a lot @anushka-singh  for pointing this and explanation about the potential issue. 

Let me check what is available in the ariadne lib for fetching this information.

Can you please point me where is this already used in code ?

Reply:
Implemented a parser and used graphQL utilities method to fetch the type of graphQL